fb.c

```diff
@@ -76,37 +77,85 @@ static ssize_t vcam_fb_write(struct fb_info *info,
+        (((int32_t) jiffies - buf->jiffies) / HZ)) {
         pr_debug("Reseting jiffies, difference %d\n",
                  ((int32_t) jiffies - buf->jiffies));
         buf->filled = 0;
+        buf->xbar = 0;
+        buf->ybar = 0;
     }
     buf->jiffies = jiffies;
 
     /* Fill the buffer */
-    to_be_copyied = length;
-    if ((buf->filled + to_be_copyied) > waiting_bytes)
-        to_be_copyied = waiting_bytes - buf->filled;
+    copy_start = 0;
+    to_be_copied = length;
 
     data = buf->data;
     if (!data) {
         pr_err("NULL pointer to framebuffer");
         return 0;
     }
 
-    if (copy_from_user(data + buf->filled, (void __user *) buffer,
-                       to_be_copyied) != 0) {
-        pr_warn("Failed to copy_from_user!");
-    }
-    buf->filled += to_be_copyied;
-
-    if (buf->filled == waiting_bytes) {
-        spin_lock_irqsave(&dev->in_q_slock, flags);
-        swap_in_queue_buffers(in_q);
-        spin_unlock_irqrestore(&dev->in_q_slock, flags);
+    if (dev->input_format.pixelformat == V4L2_PIX_FMT_RGB24) {
+        size_t bytesperpixel = 3;
+        size_t line_vir = info->var.xres_virtual * bytesperpixel;
+        size_t y_vir = info->var.yres_virtual;
+        size_t line_min = info->var.xoffset * bytesperpixel;
+        size_t y_min = info->var.yoffset;
+        size_t line_max = (info->var.xoffset + info->var.xres) * bytesperpixel;
+        size_t y_max = (info->var.yoffset + info->var.yres);
+
+        while (to_be_copied > 0 && buf->ybar < y_vir) {
+            if (buf->ybar < y_min || buf->ybar >= y_max ||
+                buf->xbar >= line_max) {
+                size_t remain = line_vir - buf->xbar;
+                if (remain > to_be_copied) {
+                    buf->xbar += to_be_copied;
+                    break;
+                } else {
+                    copy_start += remain;
+                    to_be_copied -= remain;
+                    buf->xbar = 0;
+                    buf->ybar += 1;
+                }
+            } else {
+                if (buf->xbar < line_min) {
+                    size_t abandon = min(line_min - buf->xbar, to_be_copied);
+                    copy_start += abandon;
+                    to_be_copied -= abandon;
+                    buf->xbar += abandon;
+                } else {
+                    size_t copyline = min(line_max - buf->xbar, to_be_copied);
+                    if (copy_from_user(data + buf->filled,
+                                       (void __user *) (buffer + copy_start),
+                                       copyline) != 0) {
+                        pr_warn("Failed to copy_from_user!");
+                    }
+                    copy_start += copyline;
+                    to_be_copied -= copyline;
+                    buf->filled += copyline;
+                    buf->xbar += copyline;
+                }
+            }
+        }
+        /* After copied, check if buf->xbar reach the border and need to carry.
```

Comment:
Improve the writing. Should be "After data is copied"
reach -> reaches.
need -> needs.